### DISCUSSION TITLE
Importing actions

### DISCUSSION THREAD

**Discussion by xixa** [statelyai/xstate#3152] [Q&A]
Hi!

So, I've been using  XState for a couple of years now and the way I ended up organizing my code is to have the actions for a given machine in a file as export functions which I then retrieve with `import * as actions from './actionsFile'` and pass it to `createMachine`. I actually do that with services and guards as well, so in the end of my machine definition I just have.

```
createMachine(machineConfig, {actions, services, guards})
```

I really enjoy this pattern, it helps keeping the length of the files manageable, the machines easier to follow and the actions reusable. Problem is, it's now completely broken on 4.30. I now get:
```
Type error: Type 'typeof import("/actionsile)' is not assignable to type 'MachineOptionsActions<TContext, TypegenDisabled & { missingImplementations: { actions: never; delays: never; guards: never; services: never; }; } & AllowAllEvents & { ...; }, Record<...>, Record<...> & { ...; }, IndexByType<...>>'.
```

I spent the weekend trying to come with a solution, but I don't seem to be able to cast `actions` to something that would satisfy the compiler. `MachineOptionsActions` is not available to import.

I would appreciate some ideas on how to solve that but I'm also interested to know if the pattern I just described is discouraged for whatever reason.

Thanks!

**Comment by davidkpiano** [comment]
Can you replicate this in a TS CodeSandbox? cc. @Andarist

**Reply by Andarist** [reply]
Ye, having a repro case for this would be very useful because the very basic example seems to work: https://codesandbox.io/s/laughing-shape-lf14v3?file=/src/index.ts:108-115 .

**Reply by xixa** [reply]
Oh, what I failed to mention is that I get that error during build (using Next.js) but it works fine on dev.
I don't know how to show that on codesandbox yet, but I'll work on it!

**Reply by Andarist** [reply]
You could setup a repo showcasing the problem then
